File: examples/xor_cipher.c
#include <string.h>
#include <stdlib.h>

char *
string_crypt(const char *input, int len, const char *key)
{
  char *output;
  int i, n;

  if(input == NULL)
    return NULL;

  output = malloc(len+1);
  output[len] = '\0';

  for(i=0, n=0; i<len; i++, n++) {
    if(key[n] == '\0')
      n = 0;
    output[i] = input[i] ^ key[n];
  }

  return output;
}

void
string_crypt_free(char *output)
{
  if(output != NULL)
    free(output);
}

File: examples/xor_cipher.pl
use strict;
use warnings;
use FFI::Platypus 2.00;
use FFI::Platypus::Buffer qw( buffer_to_scalar );
use YAML ();

my $ffi = FFI::Platypus->new(
  api => 2,
  lib => './xor_cipher.so',
);

$ffi->attach( string_crypt_free => ['opaque'] );

$ffi->attach( string_crypt => ['string','int','string'] => 'opaque' => sub{
  my($xsub, $input, $key) = @_;
  my $ptr = $xsub->($input, length($input), $key);
  my $output = buffer_to_scalar $ptr, length($input);
  string_crypt_free($ptr);
  return $output;
});

my $orig = "hello world";
my $key  = "foobar";

print YAML::Dump($orig);
my $encrypted = string_crypt($orig, $key);
print YAML::Dump($encrypted);
my $decrypted = string_crypt($encrypted, $key);
print YAML::Dump($decrypted);
